### PR TITLE
[REG-1475] Support raw key presses in bots

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Adds a new `RGAction_KeyPress` class that supports raw key input events using Unity's InputSystem for bots
- Updates RGGizmos to support Vector3 offsets instead of just yOffsets <- **THIS IS NOT BACKWARDS COMPATIBLE**
  - This will currently only impact other RG developers and their example bots (like @vontell racer bot)
- Updates RGGizmos to be usable by our SDK internals, not just user bot code.
- Discussion Points/Todos
  - This currently REQUIRES inclusion of Unity's InputSystem as a dependency, thus limiting our supported customer set.  I need help finding a way to restructure our unity package (and repo) so that we can easily package stuff like this as 'additional downloads'
    - I found a limitation of manually created RGAction classes in our generation code for agent builder.  We handled this pretty well for states, but need some work for actions to make this fully generic.  See [(Linear REG-1476)](https://linear.app/regression-games/issue/REG-1476/unity-sdk-find-hand-coded-rgaction-classes-in-ab-extract)
  - Documentation additions/updates about using this in bots
- Bot Code Example
```csharp
        private long _millis = 0;

        public override void ProcessTick(RG rgObject)
        {
            //Push a random key every second and hold it for 2 seconds
            long milliseconds = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;
            if (milliseconds - _millis > 1000)
            {
                _millis = milliseconds;

                var thisEntity = rgObject.FindPlayers().First();
                
                try
                {
                    Key[] keys = { Key.W, Key.A, Key.S, Key.D };
                    var random = new System.Random();
                    var index = random.Next(0, keys.Length);
                    var action = new RGActionRequest_KeyPress(keys[index], 2.0);
                    rgObject.PerformAction(action);
                }
                catch (Exception ex)
                {
                    RGDebug.LogError($"Error pressing key: {ex}");
                }
            }
        }
```

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

**Comment by RG-nAmKcAz** [comment]
@vontell @rcornwall-reg 

YES, this is a draft PR that I'm asking for review on.  It's not 'DONE' because I need help/input on how to package this correctly so that our core SDK doesn't rely on Unity's InputSystem.

File: Runtime/Scripts/DebugUtils/BillboardText.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@ public class BillboardText : MonoBehaviour
 {
     private TextMeshProUGUI _text;
     public string content = "";
-    public float yOffset = 2f;
+    public Vector3 offset = new Vector3(0f, 2f, 0f);
```
Comment (RG-nAmKcAz) [inline Runtime/Scripts/DebugUtils/BillboardText.cs:9]: I needed z offset to handle some object overlap/render depth issues.

x offset was also useful for some objects depending on their location on the screen

File: Runtime/Scripts/DebugUtils/RGGizmos.cs [hunk 2]
```diff
@@ -289,7 +291,7 @@ public void DestroyAllSpheres()
          * Draws all Gizmos that have been set.
          * </summary>
          */
-        protected internal void OnDrawGizmos()
+        public void OnDrawGizmos()
```
Comment (RG-nAmKcAz) [inline Runtime/Scripts/DebugUtils/RGGizmos.cs:294]: I'm not thrilled with this change, but it was the cleanest backwards compatible solution I found for being able to call this internally from our SDK classes, not just user bots.

File: Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using RegressionGames.StateActionTypes;
+using RegressionGames.DebugUtils;
+using UnityEngine;
+#if ENABLE_INPUT_SYSTEM
```
Comment (RG-nAmKcAz) [inline Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs:8]: These currently don't do a whole lot...

Ideally, we can package this class as a separate thing to add on to our SDK so you only pull in dependencies on Unity's inputSystem when you really need to.

**Review by vontell — APPROVED** [review]
TBH... I say we don't even care that we have the input system included. Might be a premature optimization - I say we just do this and if we get a customer that can't have it, we handle it then.

File: Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                }
+
+                // force updating these now
+                RgGizmos.OnDrawGizmos();
+            }
+            else
+            {
+                RgGizmos.DestroyText(id);
+                // force updating these now
+                RgGizmos.OnDrawGizmos();
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs:143]: I could move this call into OnDrawGizmos in this same class I suppose.  That would be more correct

File: Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            
+            // handle un-pressing any keys that have expired their time
+            var keys = keysDown.Keys.ToList();
+            foreach (var key in keys)
+            {
+                // this is < instead of <= so that keys last at least 1 frame
+                if (keysDown[key] < Time.unscaledTime)
+                {
+                    // dictionary will have this value because we clicked it
+                    UnPressKey(key, _inputActions.GetValueOrDefault(key));
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs:104]: That is the intent yes, which is why I keep updating the end time for that key to be release in the tracking map.  In PressKey, it doesn't send the new key down event to the input system if the key is already down... and the end time to send key up is updated accordingly.

I do apparently have a bug/deficiency in the inverse of your case though.. i'll need to fix this

What if I have the following:
t = 0, start holding "A" for 1 second
t = 0.5, start holding "A" for 0.2 second
 --- "A" currently gets released at 0.7 instead of 1.0 with the current code.
 
 ```csharp
 // set or update the un-press time tracker
var timeToUnPress = Time.unscaledTime + (holdTime > 0f ? holdTime : 0f);
// TODO: NEED CODE HERE TO VALIDATE THAT NEW VALUE IS AFTER PREVIOUS VALUE
keysDown[key] = timeToUnPress;
 ```

File: Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (vontell) [reply Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs:143]: Yeah you'd basically just want to move the RgGizmos.OnDrawGizmos(); call to the monob's OnDrawGizmos functon, but keep everything else in Update since OnDrawGizmos is only called when the Gizmo feature is turned on.

**Review by vontell — APPROVED** [review]

File: Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotConfigs/RGAction_KeyPress.cs:104]: @vontell ... Realized I never pushed this... now pushed in this commit
https://github.com/Regression-Games/RGUnityBots/pull/104/commits/bf9ec863a9b241ac21001e99dccf8770cb800359

**Review by vontell — APPROVED** [review]
